### PR TITLE
WS2812 API rework

### PR DESCRIPTION
**Pull request by fauxpark**
## Description

Some pretty huge changes to the WS2812 API in order to clean a few things up, in particular extrication of the WS2812-specific byte ordering from `rgb_led_t`.

The `ws2812_setleds()` API has been replaced with the more RGB Matrix-style `init/set_color/set_color_all/flush` system, with a new type `ws2812_led_t` and array of such in each driver to hold the pixel data. The `init` function has already been implemented, so not much to do there. `set_color()` and `set_color_all()` simply assign the RGB components to the given index(es) in the array. `flush()` mostly contains the existing `setleds()` functionality, but no longer takes the LED array and LED count as arguments.

Since this is a WIP, a bunch of things will now be broken:
- ~~The LED index adjustment for split boards in `rgb_matrix_drivers.c` has been removed for the moment, so split WS2812 likely no longer works~~
- ~~RGBLight (Underglow) has not been touched yet, this code won't compile as it relies on the old API~~

Testing progress (RGB Matrix):

|Driver |Channels|Arch|Board       |Status                  |
|-------|--------|----|------------|------------------------|
|Bitbang|RGB     |AVR |Pro Micro   |:heavy_check_mark:      |
|Bitbang|RGBW    |AVR |Pro Micro   |:heavy_check_mark:      |
|I2C    |RGB     |AVR |JJ4x4       |:heavy_check_mark:      |
|Bitbang|RGB     |ARM |Proton-C    |:heavy_check_mark:      |
|Bitbang|RGBW    |ARM |Proton-C    |:heavy_check_mark:      |
|PIO    |RGB     |ARM |RP2040 Pico |:heavy_check_mark:      |
|PIO    |RGBW    |ARM |RP2040 Pico |:heavy_check_mark:      |
|PWM    |RGB     |ARM |Proton-C    |:heavy_check_mark:      |
|PWM    |RGBW    |ARM |Proton-C    |:heavy_check_mark:      |
|SPI    |RGB     |ARM |Proton-C    |:heavy_check_mark:      |
|SPI    |RGBW    |ARM |Proton-C    |:heavy_check_mark:      |

Testing progress (RGBLight):
 * Animations :heavy_check_mark:
 * Index Remapping :heavy_check_mark:
 * Lighting Layers :heavy_check_mark:
 * Clipping Ranges :heavy_check_mark:
 * Split :heavy_check_mark:
 * Velocikey :heavy_check_mark:

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: platforms/avr/drivers/ws2812_bitbang.c [hunk 1]
```diff
@@ -172,3 +150,33 @@ static inline void ws2812_sendarray_mask(uint8_t *data, uint16_t datlen, uint8_t
 
     SREG = sreg_prev;
 }
+
+ws2812_led_t ws2812_leds[WS2812_LED_COUNT];
```
Comment (zvecr) [inline platforms/avr/drivers/ws2812_bitbang.c:154]: This is potentially going to cause issue as there are already keyboards that overflow the available ram. Worse that you generally only find out at runtime.

Comment (fauxpark) [reply platforms/avr/drivers/ws2812_bitbang.c:154]: RGB Matrix already had an `rgb_led_t` array in the WS2812 shim, and RGBLight currently contains its own internal array which it operates on, and which will be removed when I get to fixing up that side. So I don't think there will be any significant increase in RAM usage, at least not from the addition of this array.

Comment (zvecr) [reply platforms/avr/drivers/ws2812_bitbang.c:154]: > So I don't think there will be any significant increase in RAM usage, at least not from the addition of this array.

So the RGBLight buffer is going to be removed as part of this PR?

Comment (fauxpark) [reply platforms/avr/drivers/ws2812_bitbang.c:154]: Yes, I'm only testing with RGB Matrix for now just to verify the driver changes work. Refactoring RGBLight to use the new driver API will be a separate part of this PR.

**Comment by fauxpark** [comment]
@daskygit Could you please test the Steelseries driver changes? And in future, avoid including quantum.h where it is not directly required.

File: keyboards/steelseries/prime_plus/rgblight_custom.c [hunk 2]
```diff
@@ -1,45 +1,65 @@
...
+    pwmCFG.channels[RGB_RED_PWM_CHANNEL - 1].mode   = PWM_OUTPUT_ACTIVE_HIGH;
+    pwmCFG.channels[RGB_GREEN_PWM_CHANNEL - 1].mode = PWM_OUTPUT_ACTIVE_HIGH;
+    pwmCFG.channels[RGB_BLUE_PWM_CHANNEL - 1].mode  = PWM_OUTPUT_ACTIVE_HIGH;
+    pwmStart(&RGB_PWM_DRIVER, &pwmCFG);
+}
+
+void set_color_custom(int index, uint8_t red, uint8_t green, uint8_t blue) {
+    prime_leds[index].r = red;
+    prime_leds[index].r = green;
+    prime_leds[index].r = blue;
```
Comment (daskygit) [inline keyboards/steelseries/prime_plus/rgblight_custom.c:30]:
```suggestion
    prime_leds[index].r = red;
    prime_leds[index].g = green;
    prime_leds[index].b = blue;
```

**Review by tzarc — APPROVED** [review]

**Pull request merged**